soccer/planning/Trajectory.hpp

```diff
@@ -0,0 +1,422 @@
+     */
+    static bool nearly_equal(const Trajectory& a, const Trajectory& b,
+                             double tolerance = 1e-6);
+
+    /**
+     * @brief Get the first instant in the past, or crash if the path is empty.
+     * @return The first instant in the path.
+     */
+    RobotInstant& first() {
+        if (empty()) {
+            throw std::runtime_error(
+                "Can't get the first() instant in an empty trajectory");
+        }
+        return instants_.front();
+    }
+
+    /**
+     * @copydoc Trajectory::first()
+     */
+    [[nodiscard]] const RobotInstant& first() const {
+        if (empty()) {
+            throw std::runtime_error(
+                "Can't get the first() instant in an empty trajectory");
+        }
+        return instants_.front();
+    }
+
+    /**
+     * Get the last instant in the past, or crash if the path is empty.
+     * @return The last instant in the path.
+     */
+    RobotInstant& last() {
+        if (empty()) {
+            throw std::runtime_error(
+                "Can't get the last() instant in an empty trajectory");
+        }
+        return instants_.back();
+    }
+
+    [[nodiscard]] const RobotInstant& last() const {
+        if (empty()) {
+            throw std::runtime_error(
+                "Can't get the last() instant in an empty trajectory");
+        }
+        return instants_.back();
+    }
+
+    /**
+     * Draw this trajectory.
+     * @param drawer The debug drawer to use.
+     * @param alt_text_position Optional position at which to draw the text in
+     * the event that the trajectory is empty.
+     */
+    void draw(DebugDrawer* drawer,
+              std::optional<Geometry2d::Point> alt_text_position =
+                  std::nullopt) const;
+
+    /**
+     * Interpolate between two robot instants
+     * @param prev previous robot instant
+     * @param next next robot instant
+     * @param time time between previous and next
+     * @return the interpolated robot instant at the given time
+     */
+    static RobotInstant interpolatedInstant(const RobotInstant& prev,
+                                            const RobotInstant& next,
+                                            RJ::Time time);
+
+    void setDebugText(std::string str) { debug_text_ = std::move(str); };
+    [[nodiscard]] std::string getDebugText() const {
+        return debug_text_ ? *debug_text_ : "";
+    }
+
+    /**
+     * @brief Get the time this trajectory was created, or nullopt if it has not
+     * yet been stamped.
+     */
+    std::optional<RJ::Time> timeCreated() { return creation_stamp_; }
+
+    /**
+     * @brief Stamp this trajectory for completion at the specified time.
+     */
+    void stamp(RJ::Time time) { creation_stamp_ = time; }
+
+    /*
+     * @brief Allows seeking to arbitrary positions in a Trajectory and
+     *  incrementing by arbitrary steps.
+     */
+    class Cursor {
```

Comment:
This header and source file are pretty large, would it be simple to split them up?

Reply:
It would be a bit annoying, `Cursor` and `Trajectory::Cursor` are pretty much interdependent.

Reply:
Nevermind then